### DISCUSSION TITLE
InputScalar / Unchanged value

### DISCUSSION THREAD

**Discussion by SadE54** [ocornut/imgui#4659] [New Users Build/Link/Run issues ONLY!]
Hi , 

I'm using  ImGui::InputScalar for some value editing in a popup : 
```cpp
if (ImGui::InputScalar("", ImGuiDataType_U8, &value, NULL, NULL, "%u", ImGuiInputTextFlags_EnterReturnsTrue)
 {
...
}
```
and currently the condition is fulfilled when the value is changed AND Enter key is pressed.
Is it possible to validate only when Enter is pressed (an possible no value change so) ? 
I looked in ImGuiInputTextFlags_ without finding a corresponding flag  🤔

**Comment by ocornut** [comment]
> ImGui::InputScalar("")

Your identifier is very likely invalid. Please read the FAQ.

> and currently the condition is fulfilled when the value is changed AND Enter key is pressed.
> Is it possible to validate only when Enter is pressed (an possible no value change so) ?

This is discussed in #701 and currently not possible. We have a InputText refactor lined up soonish which might facilitate further improvements.

**Comment by SadE54** [comment]
Ok thanks

I tried to add a 'Cancel' button to allow closing popup in such a case , but I think I have the same problem than this one : [#134](https://github.com/ocornut/imgui/issues/134)


![explorer_1Evs6j6Mmi](https://user-images.githubusercontent.com/2878768/137720227-29e43988-ab7f-43e9-863c-822d788c40d2.png)

If I'm using ImGui::SetKeyboardFocusHere(); for my input field, to allow direct focus when opening the popup , I cannot press the 'Cancel' button to close the window 😆
I tried the fix mentioned without success : 
```cpp
if (ImGui::GetWindowIsFocused() && !ImGui::IsAnyItemActive()) 
    ImGui::SetKeyboardFocusHere();
```

**Reply by ocornut** [reply]
Well that code is wrong, the code doesn't match the intent of your sentence "when opening the popup".
Maybe you should use `IsWindowAppearing()`.

**Reply by SadE54** [reply]
I was just playing with IsWindowAppearing when you answered , but without success . It looks like there's a timing problem, like if it was too soon to set up the focus when IsWindowAppearing() return true
If I set the focus with a button directly , it's working well. Here's the sample code:

```cpp
static void ShowParameterEdit(bool* p_open)
{
    ImGui::OpenPopup("Info");
    if (ImGui::BeginPopup("Edit", ImGuiWindowFlags_AlwaysAutoResize))
    {
        ImGui::Text("Edit value");
        ImGui::Separator();

        bool focus = false;
		
	// Using the button , focus can be set correctly on input scalar
        if (ImGui::Button("Focus"))
        {
            focus = true;
        } 
		
	// The following block is called on when popup is opening, but focus is not set on input scalar
        if (ImGui::IsWindowAppearing())
        {
            focus = true;
        }

	    if (focus)
		{
			ImGui::SetKeyboardFocusHere();
		}
        
        float value = 0.0f;
        ImGui::InputScalar("##float_edit", ImGuiDataType_Float, &value, NULL, NULL, "%f", ImGuiInputTextFlags_EnterReturnsTrue);

		if (ImGui::Button("Cancel"))
        {
            ImGui::CloseCurrentPopup();
            *p_open = false;
        }

        ImGui::EndPopup();
	}
}
```

**Reply by ocornut** [reply]
There was a bug with `SetKeyboardFocusHere()` on clipped item fixed recently in 1.85 but we don't know what version number you are using.

**Reply by SadE54** [reply]
I'm using 1.84 , sorry .
I tested playing with a counter for  each frame after IsWindowAppearing is triggered.
A correct focus is done the frame just after .

**Reply by ocornut** [reply]
If you update 1.85 this will be fixed.

**Reply by SadE54** [reply]
Many thanks for your time and help 👌If sometimes I can feel myself a little lost over all features, It's a real pleasure for me to work with such a tool , especially when you're used to cumbersome wpf 😁
